File: services/r/graphvis_plugin.R
#!/usr/bin/env Rscript
# Optional R scientific-service example. Requires jsonlite + arrow.
suppressPackageStartupMessages(library(jsonlite))
suppressPackageStartupMessages(library(arrow))
con <- file("stdin", open="r")
while(length(line <- readLines(con, n=1, warn=FALSE)) > 0) {
  result <- tryCatch({
    req <- fromJSON(line); op <- ifelse(is.null(req$op), "health", req$op)
    if(op == "health") list(ok=TRUE, service="graphvis-r-science", version="18.4.0", transport="Arrow IPC + JSON control")
    else if(op == "column.mean") { tab <- read_ipc_file(req$arrow_path); list(ok=TRUE, result=mean(tab[[req$column]], na.rm=TRUE)) }
    else list(ok=FALSE, error=paste("Unknown operation:", op))
  }, error=function(e) list(ok=FALSE,error=conditionMessage(e)))
  cat(toJSON(result,auto_unbox=TRUE),"\n",sep=""); flush.console()
}

File: services/julia/GraphVisPlugin.jl
#!/usr/bin/env julia
# Optional Julia scientific-service example. Bulk arrays use Arrow IPC.
using JSON3
try
    using Arrow
catch
    println(stderr, "Install Arrow.jl and JSON3.jl to use the GraphVis Julia service")
end
for line in eachline(stdin)
    try
        req=JSON3.read(line)
        op=String(get(req,:op,"health"))
        if op=="health"
            println(JSON3.write((ok=true,service="graphvis-julia-science",version="18.4.0",transport="Arrow IPC + JSON control")))
        elseif op=="column.mean"
            tbl=Arrow.Table(String(req[:arrow_path])); col=getproperty(tbl,Symbol(String(req[:column]))); println(JSON3.write((ok=true,result=sum(skipmissing(col))/length(collect(skipmissing(col))))))
        else
            println(JSON3.write((ok=false,error="Unknown operation: $op")))
        end
    catch e
        println(JSON3.write((ok=false,error=string(e))))
    end
    flush(stdout)
end
